jsettlers.logic/src/jsettlers/ai/highlevel/AiStatistics.java

```diff
@@ -0,0 +1,612 @@
+		if (nearestRightPoint != null) {
+			currentNearestPointDistance = point.calculateDistanceTo(nearestRightPoint);
+			result = nearestRightPoint;
+		}
+		ShortPoint2D nearestLeftPoint = getNearestPoinInDefaultPartionOutOfSortedMapInXDirection(point, sortedPoints, currentNearestPointDistance,
+				new Integer(point.x - 1), -1, -1, playerId);
+		if (nearestLeftPoint != null) {
+			result = nearestLeftPoint;
+		}
+		return result;
+	}
+
+	private ShortPoint2D getNearestPoinInDefaultPartionOutOfSortedMapInXDirection(ShortPoint2D point, Map<Integer, List<Integer>> sortedPoints,
+			double currentNearestPointDistance, Integer x, Integer increment, Integer border, byte playerId) {
+		if (x.equals(border) || Math.abs(x - point.x) > currentNearestPointDistance) {
+			return null;
+		}
+		if (!sortedPoints.containsKey(x)) {
+			return getNearestPoinInDefaultPartionOutOfSortedMapInXDirection(point, sortedPoints, currentNearestPointDistance, x + increment,
+					increment, border, playerId);
+		}
+		ShortPoint2D result = null;
+		ShortPoint2D southYPoint = getNearestPoinInDefaultPartitionOutOfSortedMapInYDirection(sortedPoints.get(x), point,
+				currentNearestPointDistance, x, new Integer(point.y), 1, new Integer(mainGrid.getHeight() + 1), playerId);
+		if (southYPoint != null) {
+			result = southYPoint;
+			currentNearestPointDistance = point.calculateDistanceTo(southYPoint);
+		}
+		ShortPoint2D northYPoint = getNearestPoinInDefaultPartitionOutOfSortedMapInYDirection(sortedPoints.get(x), point,
+				currentNearestPointDistance, x, new Integer(point.y - 1), -1, -1, playerId);
+		if (northYPoint != null) {
+			result = northYPoint;
+			currentNearestPointDistance = point.calculateDistanceTo(northYPoint);
+		}
+		if (Math.abs(point.x - (x + increment)) < currentNearestPointDistance) {
+			ShortPoint2D nextPoint = getNearestPoinInDefaultPartionOutOfSortedMapInXDirection(point, sortedPoints, currentNearestPointDistance, x
+					+ increment, increment, border, playerId);
+			if (nextPoint != null) {
+				result = nextPoint;
+			}
+		}
+		return result;
+	}
+
+	private ShortPoint2D getNearestPoinInDefaultPartitionOutOfSortedMapInYDirection(List<Integer> ypsilons, ShortPoint2D point,
+			double currentNearestPointDistance, Integer x, Integer y, Integer increment, Integer border, byte playerId) {
+		if (y.equals(border) || Math.abs(y - point.y) > currentNearestPointDistance) {
+			return null;
+		}
+		if (!ypsilons.contains(y) || partitionsGrid.getPartitionAt(x, y).getPlayerId() != playerId) {
+			return getNearestPoinInDefaultPartitionOutOfSortedMapInYDirection(ypsilons, point, currentNearestPointDistance, x, y + increment,
+					increment, border, playerId);
+		}
+		if (point.calculateDistanceTo(new ShortPoint2D(x, y)) > currentNearestPointDistance) {
+			return null;
+		}
+		return new ShortPoint2D(x, y);
+	}
+
+	public List<ShortPoint2D> getMovablePositionsByTypeForPlayer(EMovableType movableType, byte playerId) {
```

Comment:
I suggest you use IMapArea as return type here. You can write a simple wrapper class for your list.